### PR TITLE
begin/copy: scan for arm executables, update arch label when found

### PR DESCRIPTION
**Pull request by dgonyeo**
The runtime.GOARCH value isn't accepted by the AppC spec as a value for
the arch label in a manifest when on an arm platform. This commit adds a
case to use "armv7l" whenever runtime.GOARCH is "arm". Additionally
(when on arm) this change scans all files at the build start and any
files copied in later for executables, and will switch the label between
"armv7l" and "armv7b" when it finds a little or big endian binary.

Fixes https://github.com/appc/acbuild/issues/190.

### CODE REVIEW

**Comment by dgonyeo** [comment]
@godart: would you be willing to try out this modified version of acbuild, and see if this fixes your issue?

**Comment by godart** [comment]
Thanks @dgonyeo. It works for me now.

I was trying myself to use the golang.org/x/sys/unix.Uname() function to determine the current architecture, but it did not compile due to bugs in golang.

**Comment by anguslees** [comment]
Hrm - is it important that this works for cross-compile situations?  If we can assume we're running natively, then it should be easy to work out current endianness and even armv6/armv7 without walking the binaries - using logic like [this xgetgoarm() function](https://github.com/golang/go/blob/master/src/cmd/dist/util.go#L516)

I feel like at some point we just have to add this as a command line option - particularly when cross-compiling - and at that point it doesn't seem worthwhile to try to be as clever as you're being here.  I can't see how we can assert generations of arm architecture other than by just trusting the user to tell us how they compiled the binaries contained in this rootfs :/  (other than completely disassembling all the code we find, looking for armv7 constructs, and assuming we found it all reliably..)

TL;DR: Can we just have a command line option, and default to something based on querying the local CPU (via /proc/cpuinfo or similar)?

File: lib/begin.go [hunk 1]
```diff
@@ -188,6 +189,22 @@ func (a *ACBuild) writeEmptyManifest() error {
 		return err
 	}
 
+	archvalue := runtime.GOARCH
+	if runtime.GOOS == "linux" && archvalue == "arm" {
+		archvalue = "armv7l"
+		walker := func(path string, info os.FileInfo, err error) error {
```
Comment (anguslees) [inline lib/begin.go]:
```go
var x uint32 = 0x01020304
switch *(*byte)(unsafe.Pointer(&x)) {
case 0x01:
    archvalue = "armv7b"
case 0x04:
    archvalue = "armv7l"
}
```

**Comment by dgonyeo** [comment]
No matter what arch label acbuild tries to specify, the user can always override it with a manual `acbuild label add arch <value>`, so I think acbuild should work fine in the case of cross-compilation.

File: lib/begin.go (hunk 1 above)
Comment (dgonyeo) [inline lib/begin.go]: What's this doing? Looking at the acbuild binary that's in memory for some specific value?

Comment (anguslees) [inline lib/begin.go]: Oh, it's just the usual way to detect local endianness - put a known multibyte value into memory, and then see how the compiler decided to actually store it.  In this case it's seeing if 0x01020304 was stored with 0x01 or 0x04 first.

Comment (anguslees) [inline lib/begin.go]: Another (better?) option might be to just use `uname -m` (aka `syscall.Uname` in golang).

Comment (s-urbaniak) [inline lib/begin.go]: We used the same trick in mesos-dns: https://github.com/mesosphere/mesos-dns/blob/883e8ace58008d18f35bca878175212eea5b6b1f/detect/masters.go#L107

Maybe we could factor this in a library function returning `binary.ByteOrder`?

**Comment by dgonyeo** [comment]
Just changed this to be more sane, now it'll just look at the endinanness of `/proc/self/exe` (based on the elf data) and pick an armv7 label. Users wishing for an armv6 label can just manually set it.

@godart would you be willing to try out this new set of changes?

File: lib/begin.go [hunk 2]
```diff
@@ -196,14 +197,31 @@ func (a *ACBuild) writeEmptyManifest() error {
 		return err
 	}
 
+	archvalue := runtime.GOARCH
+	if runtime.GOOS == "linux" && archvalue == "arm" {
+		elfFile, err := elf.Open("/proc/self/exe")
```
Comment (anguslees) [inline lib/begin.go]: .. so why do this rather than my 0x01020304 approach suggested [earlier](https://github.com/appc/acbuild/pull/195#discussion_r58650830)?

Comment (dgonyeo) [inline lib/begin.go]: Mostly because this was already written :p. I'll switch it over to use your method tomorrow morning.

**Comment by dgonyeo** [comment]
Just pushed the change to use @anguslees's method for determining endianness.

**Comment by anguslees** [comment]
lgtm (fwiw)

**Comment by dgonyeo** [comment]
It would be real convenient if I had a couple of arm devices to test this on :/

**Comment by jonboulle** [comment]
did you fill out the lab survey? ;-)

**Comment by dgonyeo** [comment]
Just pushed changes to also handle aarch64 and aarch64_be, and tested it out on an aarch64 machine courtesy of redbeard. Looks like it works fine at least in that case.

**Comment by jonboulle** [comment]
LGTM

**Comment by godart** [comment]
Sorry. I've been away from my account for a while.
Works well for my Cubietruck (armv7l) as well.
